gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/FsBenchmark.java

```diff
@@ -0,0 +1,468 @@
+                fs, testFile, /* readSize= */ 1024, /* numReads= */ 1, /* numThreads= */ 2));
+
+    benchmarkRead(
+        fs,
+        testFile,
+        parseInt(args.getOrDefault("--read-size", String.valueOf(1024))),
+        parseInt(args.getOrDefault("--num-reads", String.valueOf(1))),
+        parseInt(args.getOrDefault("--num-threads", String.valueOf(1))));
+
+    return 0;
+  }
+
+  private void benchmarkRead(
+      FileSystem fs, Path testFile, int readSize, int numReads, int numThreads) {
+    System.out.printf(
+        "Running read test using %d bytes reads to fully read '%s' file %d times in %d threads%n",
+        readSize, testFile, numReads, numThreads);
+
+    List<LongSummaryStatistics> readFileBytesList = new ArrayList<>(numThreads);
+    List<LongSummaryStatistics> readFileTimeNsList = new ArrayList<>(numThreads);
+
+    List<LongSummaryStatistics> readCallBytesList = new ArrayList<>(numThreads);
+    List<LongSummaryStatistics> readCallTimeNsList = new ArrayList<>(numThreads);
+
+    ExecutorService executor = Executors.newFixedThreadPool(numThreads);
+    CountDownLatch initLatch = new CountDownLatch(numThreads);
+    CountDownLatch startLatch = new CountDownLatch(1);
+    CountDownLatch stopLatch = new CountDownLatch(numThreads);
+    List<Future<?>> futures = new ArrayList<>(numThreads);
+    for (int i = 0; i < numThreads; i++) {
+      futures.add(
+          executor.submit(
+              () -> {
+                LongSummaryStatistics readFileBytes = newLongSummaryStatistics(readFileBytesList);
+                LongSummaryStatistics readFileTimeNs = newLongSummaryStatistics(readFileTimeNsList);
+
+                LongSummaryStatistics readCallBytes = newLongSummaryStatistics(readCallBytesList);
+                LongSummaryStatistics readCallTimeNs = newLongSummaryStatistics(readCallTimeNsList);
+
+                byte[] readBuffer = new byte[readSize];
+
+                initLatch.countDown();
+                awaitUninterruptibly(startLatch);
+                try {
+                  for (int j = 0; j < numReads; j++) {
+                    try (FSDataInputStream input = fs.open(testFile)) {
+                      long readStart = System.nanoTime();
+                      long fileBytesRead = 0;
+                      int bytesRead;
+                      do {
+                        long readCallStart = System.nanoTime();
+                        bytesRead = input.read(readBuffer);
+                        if (bytesRead > 0) {
+                          fileBytesRead += bytesRead;
+                          readCallBytes.accept(bytesRead);
+                        }
+                        readCallTimeNs.accept(System.nanoTime() - readCallStart);
+                      } while (bytesRead >= 0);
+
+                      readFileBytes.accept(fileBytesRead);
+                      readFileTimeNs.accept(System.nanoTime() - readStart);
+                    }
+                  }
+                } finally {
```

Comment:
Is the try-finally block needed here? Does simply stopLatch.countDown() right after the big for loop won't work?

Reply:
Yes, without it if exception will be thrown benchmark will never terminate and will not print this exception.